music_assistant/providers/music_insights/chroma_embeddings.py

```diff
@@ -0,0 +1,280 @@
+        with torch.no_grad():
+            feats = self.model.get_text_features(**inputs.to(self.model.device))
+        return cast("np.ndarray[Any, Any]", feats.squeeze(0).cpu().numpy())
+
+    async def embeddings_from_audio(
+        self, samples: np.ndarray[Any, Any], sr: int = 48_000
+    ) -> np.ndarray[Any, Any]:
+        """
+        Generate embeddings asynchronously for given audio samples.
+
+        Args:
+            samples: A numpy array of audio samples.
+            sr: The sample rate of the audio.
+
+        Returns:
+            A numpy array representing the audio embedding.
+        """
+        return await asyncio.to_thread(self._embed_audio, samples, sr)
+
+    def _embed_audio(self, samples: np.ndarray[Any, Any], sr: int) -> np.ndarray[Any, Any]:
+        """Embed audio using the CLAP model (synchronous)."""
+        inputs = self.processor(audios=samples, sampling_rate=sr, return_tensors="pt", padding=True)
+        with torch.no_grad():
+            feats = self.model.get_audio_features(**inputs.to(self.model.device))
+        return cast("np.ndarray[Any, Any]", feats.squeeze(0).cpu().numpy())
+
+    def _window_embed(
+        self, samples: np.ndarray[Any, Any], sr: int = 48_000
+    ) -> np.ndarray[Any, Any]:
+        """
+        Embed audio using a sliding window approach and average the results.
+
+        Args:
+            samples: A numpy array of audio samples.
+            sr: The sample rate of the audio.
+
+        Returns:
+            A numpy array representing the averaged audio embedding over windows.
+        """
+        step = int(sr * self.audio_hop_s)
+        win = int(sr * self.audio_window_s)
+        if len(samples) <= win:
+            return self._embed_audio(samples, sr)
+        chunks = [
+            self._embed_audio(samples[i : i + win], sr)
+            for i in range(0, len(samples) - win + 1, step)
+        ]
+        return cast("np.ndarray[Any, Any]", np.mean(np.vstack(chunks), axis=0))
+
+    async def upsert_track(self, track: Track) -> None:
+        """
+        Generate and upsert text embeddings for a track into ChromaDB.
+
+        Combines track metadata (genres, artists, name, album, mood) into a
+        single string for text embedding. Audio embedding is currently commented out.
+
+        Args:
+            track: The Track object to process.
+        """
+        album_name = track.album.name if track.album else "Unknown Album"
+        artist_names = [a.name for a in track.artists] if track.artists else []
+        genres = ",".join(track.metadata.genres or [])
+        mood = track.metadata.mood or ""
+        text = f"{genres},{','.join(artist_names)},{track.name},{album_name},{mood}"
+        text_emb = await self.embeddings_from_str(text)
+
+        docs: list[str] = [text]
+        embs: list[np.ndarray[Any, Any]] = [text_emb]
+        ids: list[str] = [track.item_id]
+
+        # waveform = None
+        # sample_rate = None
```

Comment:
If you need help with this part, ping me on discord. Its relatively easy to get the audio stream in pcm.